### PR TITLE
Yaswanth19 add janus model

### PR DESCRIPTION
**Pull request by hsilva664**
PR to add preliminary version of the generative part of Janus. The pretrained weights are working correctly. Logits are matching.

### CODE REVIEW

File: src/transformers/models/janus/modeling_janus.py [hunk 1]
```diff
@@ -47,27 +46,542 @@
...
-    import torch.nn as nn
-    import torch.nn.functional as F
-
 logger = logging.get_logger(__name__)
 
 # General docstring
 _CONFIG_FOR_DOC = "JanusConfig"
 
 
+def janus_gen_vision_nonlinearity(x):
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:72]: We don't need this function as we can do `hidden_states*=torch.sigmoid(hidden_states)` directly

File: src/transformers/models/janus/modeling_janus.py [hunk 2]
```diff
@@ -47,27 +46,542 @@
...
+    # swish
+    return x * torch.sigmoid(x)
+
+
+def JanusGenVisionNormalize(in_channels, norm_type="group"):
+    assert norm_type in ["group", "batch"]
+    if norm_type == "group":
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:79]: Also i think every where we are using GroupNorm so we can directly intiliaze `GroupNorm` in Module __init__ as this function is redundant. We don't have to provide feature/plug-and-play in HF. Only the necessary and used ones

File: src/transformers/models/janus/modeling_janus.py [hunk 3]
```diff
@@ -47,27 +46,542 @@
...
+        return nn.GroupNorm(num_groups=32, num_channels=in_channels, eps=1e-6, affine=True)
+    elif norm_type == "batch":
+        return nn.SyncBatchNorm(in_channels)
+
+
+class JanusGenVisionEncoder(nn.Module):
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:85]: Can be renamed to `JanusVQVAEEncoder` which is more appropriate

File: src/transformers/models/janus/modeling_janus.py [hunk 4]
```diff
@@ -47,27 +46,542 @@
...
+    def __init__(
+        self,
+        in_channels=3,
+        ch=128,
+        ch_mult=(1, 1, 2, 2, 4),
+        num_res_blocks=2,
+        norm_type="group",
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:92]: In `transformers` we don't pass parameters in main modules. These all parameters need to be defined in config and only config should be passed here. Please refer to the below commit to see how I have defined those same blocks. https://github.com/huggingface/transformers/pull/36053/commits/c4dbe959a254ec3f0b3206ee5c857f8d4222d9b3

File: src/transformers/models/janus/modeling_janus.py [hunk 5]
```diff
@@ -47,27 +46,542 @@
...
+        self.mid = nn.ModuleList()
+        self.mid.append(JanusGenVisionResnetBlock(block_in, block_in, dropout=dropout, norm_type=norm_type))
+        self.mid.append(JanusGenVisionAttnBlock(block_in, norm_type=norm_type))
+        self.mid.append(JanusGenVisionResnetBlock(block_in, block_in, dropout=dropout, norm_type=norm_type))
+
+        # end
+        self.norm_out = JanusGenVisionNormalize(block_in, norm_type)
+        self.conv_out = nn.Conv2d(block_in, z_channels, kernel_size=3, stride=1, padding=1)
+
+    def forward(self, x):
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:134]: Nit: We can use more descriptive names like hidden_states, residual etc.. instead of shortforms.

File: src/transformers/models/janus/modeling_janus.py [hunk 6]
```diff
@@ -47,27 +46,542 @@
...
+            if i_level != 0:
+                conv_block.upsample = JanusGenVisionUpsample(block_in, resamp_with_conv)
+            self.conv_blocks.append(conv_block)
+
+        # end
+        self.norm_out = JanusGenVisionNormalize(block_in, norm_type)
+        self.conv_out = nn.Conv2d(block_in, out_channels, kernel_size=3, stride=1, padding=1)
+
+    @property
+    def last_layer(self):
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:207]: This property is not required ig if not used anywhere.

File: src/transformers/models/janus/modeling_janus.py [hunk 7]
```diff
@@ -47,27 +46,542 @@
...
+        min_encoding_indices = torch.argmin(d, dim=1)
+        z_q = embedding[min_encoding_indices].view(z.shape)
+        perplexity = None
+        min_encodings = None
+        vq_loss = None
+        commit_loss = None
+        entropy_loss = None
+
+        # compute loss for embedding
+        if self.training:
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:295]: This is the training related code I was talking about. But I am not sure while finetuning from scratch we will also train this Module. This can be added at the end if required.

File: src/transformers/models/janus/modeling_janus.py [hunk 8]
```diff
@@ -47,27 +46,542 @@
...
+            if channel_first:
+                z_q = z_q.reshape(shape[0], shape[2], shape[3], shape[1])
+                # reshape back to match original input shape
+                z_q = z_q.permute(0, 3, 1, 2).contiguous()
+            else:
+                z_q = z_q.view(shape)
+        return z_q
+
+
+class JanusGenVisionResnetBlock(nn.Module):
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:330]: The `Attention` and `Resnet` is similar to the one in Chameleon Model. So we can utilize the Modular feature here

File: src/transformers/models/janus/modeling_janus.py [hunk 9]
```diff
@@ -47,27 +46,542 @@
...
 )
 from ...utils.deprecation import deprecate_kwarg
-from .configuration_janus import JanusConfig, JanusTextConfig, JanusVisionConfig, JanusVQVAEConfig
+from .configuration_janus import (
+    JanusConfig,
+    JanusGenAlignerConfig,
+    JanusGenHeadConfig,
+    JanusGenVisionConfig,
+    JanusTextConfig,
+    JanusVisionConfig,
```
Comment (yaswanth19) [inline src/transformers/models/janus/modeling_janus.py:59]: AFAIK we can't have this many configs. Only `JanusVision`, `JanusText`, `JanusVQVAE` and `JanusConfig`. So we have to move the Aligner/GenHead config to relevant Module configs. I will handle that in my changes.

File: src/transformers/models/janus/modeling_janus.py (hunk 7 above)
Comment (hsilva664) [reply src/transformers/models/janus/modeling_janus.py:295]: I believe it should be kept if training from scratch or even just finetuning.
